spring-kafka/src/test/java/org/springframework/kafka/core/KafkaTemplateTests.java

```diff
@@ -195,4 +195,14 @@ public void testTemplateDisambiguation() throws Exception {
 		pf.createProducer().close();
 	}
 
+	@Test(expected = IllegalStateException.class)
+	public void flushWithoutSend() throws Exception {
+		Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+		DefaultKafkaProducerFactory<String, String> pf = new DefaultKafkaProducerFactory<String, String>(senderProps);
+		pf.setKeySerializer(new StringSerializer());
+		KafkaTemplate<String, String> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(STRING_KEY_TOPIC);
+		template.flush();
+		throw new AssertionError("IllegalStateException must be thrown upon flushing");
```

Comment:
Great! Merging with some code style polishing and your name for the affected classes.

Regarding this test-case some nit-picks. 
We don't need so much options for the `KafkaTemplate` since there is just enough to call its `flush()` and don't let any one after us to override the code without issues.
On the other hand we stopped with Gary rely on the `@Test(expected =` because there might be more places where such an exception can be thrown. That's why our pattern is like:

```
try {
      invocation_to_verify;
      fail("Some_Exception expceted");
}
catch {Exception e} {
      assertThat(e, instanceOf(Some_Exception.class));
      assertThat(e.getMessage(), containsString("Some part of the exception message to verify"));
}
```
Will fix on merge and share in this PR for future considerations.